File: android/app/src/main/java/ws/baseline/paradrone/geo/SegmentLine.java
package ws.baseline.paradrone.geo;

import androidx.annotation.NonNull;
import java.util.Arrays;
import java.util.List;
import timber.log.Timber;

import static ws.baseline.paradrone.util.Numbers.interpolate;

public class SegmentLine implements Segment {

    @NonNull
    private final PointV start;
    @NonNull
    private final PointV end;

    public SegmentLine(@NonNull Point start, @NonNull Point end) {
        final double dx = end.x - start.x;
        final double dy = end.x - start.y;
        final double len = Math.sqrt(dx * dx + dy * dy);
        this.start = new PointV(start.x, start.y, dx / len, dy / len);
        this.end = new PointV(end.x, end.y, dx / len, dy / len);
        if (len == 0) {
            Timber.e("Invalid line: same start and end %s", start);
        }
    }

    @NonNull
    @Override
    public PointV start() {
        return start;
    }

    @NonNull
    @Override
    public PointV end() {
        return end;
    }

    @NonNull
    @Override
    public Path fly(double distance) {
        if (distance < 0) {
            Timber.e("Flight distance cannot be negative %s", distance);
        }
        // Linear interpolate
        final double alpha = distance / this.length();
        final Point proj = new Point(
                interpolate(this.start.x, this.end.x, alpha),
                interpolate(this.start.y, this.end.y, alpha)
        );
        return new Path("line-fly", new SegmentLine(start, proj));
    }

    @Override
    public double length() {
        final double dx = start.x - end.x;
        final double dy = start.y - end.y;
        return Math.sqrt(dx * dx + dy * dy);
    }

    @NonNull
    @Override
    public List<Point> render() {
        return Arrays.asList(start, end);
    }

    @NonNull
    @Override
    public String toString() {
        return "Line " + length();
    }
}

File: autopilot/lib/paralib/segment_line.cpp
#include <math.h>
#include <stdio.h>
#include "path.h"

static double interpolate(double start, double end, double alpha);

PointV line_start(Line *line) {
  const double dx = line->end.x - line->start.x;
  const double dy = line->end.y - line->start.y;
  const double len = sqrt(dx * dx + dy * dy);
  return PointV {line->start.x, line->start.y, dx / len, dy / len};
}

PointV line_end(Line *line) {
  const double dx = line->end.x - line->start.x;
  const double dy = line->end.y - line->start.y;
  const double len = sqrt(dx * dx + dy * dy);
  return PointV {line->end.x, line->end.y, dx / len, dy / len};
}

double line_length(Line *line) {
  const double dx = line->start.x - line->end.x;
  const double dy = line->start.y - line->end.y;
  return sqrt(dx * dx + dy * dy);
}

/**
 * Fly a given distance along the path
 */
Path *line_fly(Line *line, const double distance) {
  if (distance < 0) {
    printf("segment_line distance must be positive %f\n", distance);
  }
  // Linear interpolate
  const double alpha = distance / line_length(line);
  Point proj = {
    interpolate(line->start.x, line->end.x, alpha),
    interpolate(line->start.y, line->end.y, alpha)
  };
  Line *extension = new Line {'L', line->start, proj};
  return new_path("line-fly", 1, (Segment**) &extension);
}

Point *line_render(Line *line) {
  return &line->start;
}

/**
 * Linear interpolation
 * @param alpha interpolation parameter from 0 (start) to 1 (end)
 */
static double interpolate(double start, double end, double alpha) {
  return start + alpha * (end - start);
}
